airflow/models/taskmixin.py

```diff
@@ -15,61 +15,76 @@
+class DependencyMixin:
+    """Mixing implementing common dependency setting methods methods like >> and <<."""
 
     @property
-    def roots(self):
-        """Should return list of root operator List[BaseOperator]"""
+    def roots(self) -> Sequence["DependencyMixin"]:
+        """
+        List of root nodes -- ones with no upstream dependencies.
+
+        a.k.a. the "start" of this sub-graph
+        """
         raise NotImplementedError()
 
     @property
-    def leaves(self):
-        """Should return list of leaf operator List[BaseOperator]"""
+    def leaves(self) -> Sequence["DependencyMixin"]:
+        """
+        List of leaf nodes -- ones with only upstream dependencies.
+
+        a.k.a. the "end" of this sub-graph
+        """
         raise NotImplementedError()
 
     @abstractmethod
-    def set_upstream(self, other: Union["TaskMixin", Sequence["TaskMixin"]]):
+    def set_upstream(self, other: Union["DependencyMixin", Sequence["DependencyMixin"]]):
         """Set a task or a task list to be directly upstream from the current task."""
         raise NotImplementedError()
 
     @abstractmethod
-    def set_downstream(self, other: Union["TaskMixin", Sequence["TaskMixin"]]):
+    def set_downstream(self, other: Union["DependencyMixin", Sequence["DependencyMixin"]]):
         """Set a task or a task list to be directly downstream from the current task."""
         raise NotImplementedError()
 
-    def update_relative(self, other: "TaskMixin", upstream=True) -> None:
+    def update_relative(self, other: "DependencyMixin", upstream=True) -> None:
         """
         Update relationship information about another TaskMixin. Default is no-op.
         Override if necessary.
         """
 
-    def __lshift__(self, other: Union["TaskMixin", Sequence["TaskMixin"]]):
+    def __lshift__(self, other: Union["DependencyMixin", Sequence["DependencyMixin"]]):
         """Implements Task << Task"""
         self.set_upstream(other)
         return other
 
-    def __rshift__(self, other: Union["TaskMixin", Sequence["TaskMixin"]]):
+    def __rshift__(self, other: Union["DependencyMixin", Sequence["DependencyMixin"]]):
         """Implements Task >> Task"""
         self.set_downstream(other)
         return other
 
-    def __rrshift__(self, other: Union["TaskMixin", Sequence["TaskMixin"]]):
+    def __rrshift__(self, other: Union["DependencyMixin", Sequence["DependencyMixin"]]):
         """Called for Task >> [Task] because list don't have __rshift__ operators."""
         self.__lshift__(other)
         return self
 
-    def __rlshift__(self, other: Union["TaskMixin", Sequence["TaskMixin"]]):
+    def __rlshift__(self, other: Union["DependencyMixin", Sequence["DependencyMixin"]]):
         """Called for Task << [Task] because list don't have __lshift__ operators."""
         self.__rshift__(other)
         return self
+
+
+class TaskMixin(DependencyMixin):
+    """:meta private:"""
+
+    def __init_subclass__(cls) -> None:
```

Comment:
curious why you only warn for subclasses?  why not warn any time this is used?

Reply:
This warns when you do `class X(TaskMixin):` -- i.e. right were you need to update the class definition. not at construction time which would point to where the sbuclass is constrcuted.

Reply:
Ah yes makes sense thank you